athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchRecordHandler.java

```diff
@@ -0,0 +1,215 @@
+                                client.getDocument(hitIterator.next()), numResultRows);
+                    }
+                    // if hitsNum < QUERY_BATCH_SIZE, then this is the last batch of documents.
+                } while (hitsNum == QUERY_BATCH_SIZE && queryStatusChecker.isQueryRunning());
+            }
+            catch (IOException error) {
+                logger.error("Error sending query:", error);
+            }
+            finally {
+                client.shutdown();
+            }
+        }
+
+        logger.info("readWithConstraint: numRows[{}] numResultRows[{}]", numRows, numResultRows.get());
+    }
+
+    /**
+     * Process the Document extracted from the SearchHit and write it's data into the BlockSpiller.
+     * @param spiller is a BlockSpiller that should be used to write the row data associated with this Document.
+     *                The BlockSpiller automatically handles chunking the response, encrypting, and spilling to S3.
+     * @param fieldIterator is the iterator for the list of schema fields extracted from the RecordRequest.
+     * @param document is the Document to be processed.
+     * @param numResultRows is the number of successfully processed Rows (Documents).
+     */
+    protected void processDocument(BlockSpiller spiller, Iterator<Field> fieldIterator,
+                                   Map<String, Object> document, AtomicLong numResultRows)
+    {
+        spiller.writeRows((Block block, int rowNum) -> {
+            boolean matched = true;
+            while (fieldIterator.hasNext()) {
+                Field field = fieldIterator.next();
+                String fieldName = field.getName();
+
+                if (!document.containsKey(fieldName)) {
+                    throw new RuntimeException("Field not found in Document: " + fieldName);
+                }
+
+                Object fieldValue = ElasticsearchHelper.coerceField(field, document.get(field.getName()));
+                Types.MinorType fieldType = Types.getMinorTypeForArrowType(field.getType());
+                try {
+                    switch (fieldType) {
+                        case LIST:
+                        case STRUCT:
+                            matched &= block.offerComplexValue(fieldName, rowNum, DEFAULT_FIELD_RESOLVER, fieldValue);
+                            break;
+                        default:
+                            matched &= block.offerValue(fieldName, rowNum, fieldValue);
+                            break;
+                    }
+                    if (!matched) {
+                        return 0;
+                    }
+                }
+                catch (Exception error) {
+                    throw new RuntimeException("Error while processing field:" + fieldName, error);
+                }
+            }
+
+            numResultRows.getAndIncrement();
+            return 1;
+        });
+    }
+
+    @VisibleForTesting
+    public int getQueryBatchSize()
```

Comment:
if this is just for testing can we make it protected? or package private?

Reply:
made protected.